Introduce minimal job worker metrics

npepinpe opened the pull request:
## Description

Adds a new API to the job worker builder that allows users to pass a custom `JobWorkerMetrics` implementation. This currently supports two metrics:

- `jobActivated(int count)`: called every time one or more jobs is activated (but not yet handled)
- `jobHandled(int count)`: called every time one or more jobs has been handled by the `JobHandler`

> **Note**
> While we currently always call the above for each individual jobs, I opted to pass in the count for future proofing.

I opted to go for a minimal set of metrics for now, on the reasoning that it's easier to add things than to remove them. From these two counters we can derive the rate of activation, the rate of handling, and subtract one from the other to estimate the amount of queued jobs.

Additionally, an optional implementation was added for Micrometer, which will help integration with Spring and Micronaut. Users can call `JobWorkerMetrics#micrometer`, which returns a builder that lets them pass a custom `MeterRegistry` and a set of fixed tags (e.g. the worker job type, the app version, etc.)

The Micrometer implementation will create the following counters:

- `zeebe.client.worker.job.activated`
- `zeebe.client.worker.job.handled`

> **Note**
> Micrometer will automatically adapt the meter names to the actual registry backend. For example, for Prometheus, the periods will be replaced with underscores, such that `zeebe.client.worker.job.activated` becomes `zeebe_client_worker_job_activated`.

Finally, some of the commits include test refactoring, namely introducing interfaces for the `JobPoller` and the `JobRunnableFactory`, as this allows us to pass controllable implementation during tests, making it very easy to unit test the job worker. This avoids doing complicated test setups with concurrency primitives to provide synchronization points (e.g. activated job but don't complete it yet) and so on.

## Related issues

closes #4700 
closes #13814 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

npepinpe commented:
If Micrometer is not on the classpath and a user calls `JobWorkerMetrics.micrometer()`, they will get the following error:

```
Exception in thread "main" java.lang.UnsupportedOperationException: Expected to create Micrometer worker metrics, but it seems Micrometer is not in your classpath
	at io.camunda.zeebe.client.api.worker.JobWorkerMetrics.micrometer(JobWorkerMetrics.java:51)
	at io.camunda.zeebe.Worker.run(Worker.java:71)
	at io.camunda.zeebe.App.createApp(App.java:50)
	at io.camunda.zeebe.Worker.main(Worker.java:114)
Caused by: java.lang.ClassNotFoundException: io.micrometer.core.instrument.MeterRegistry
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at io.camunda.zeebe.client.api.worker.JobWorkerMetrics.micrometer(JobWorkerMetrics.java:49)
	... 3 more
```

npepinpe commented on clients/java/src/main/java/io/camunda/zeebe/client/api/worker/JobWorkerMetrics.java:
```diff
@@ -0,0 +1,63 @@
+   *
+   * @param count the amount of jobs that were activated
+   */
+  default void jobActivated(final int count) {}
```
I opted to make all methods default to simplify partial implementations, i.e. not all users may care about all metrics :shrug:

npepinpe commented on clients/java/src/main/java/io/camunda/zeebe/client/api/worker/metrics/MicrometerJobWorkerMetricsBuilder.java:
```diff
@@ -0,0 +1,83 @@
+   * @param tags the tags to apply to all metrics
+   * @return this builder for chaining
+   */
+  MicrometerJobWorkerMetricsBuilder withTags(final Iterable<Tag> tags);
```
To keep things simple, I figured let's let the user decide on whether to pass the worker type as a tag or not. It's a little verbose, but since these could potentially be very high cardinality tags, I'd rather not make that decision for them.

npepinpe commented on clients/java/src/main/java/io/camunda/zeebe/client/impl/worker/JobPollerImpl.java:
```diff
@@ -0,0 +1,164 @@
+import java.util.function.IntConsumer;
+import org.slf4j.Logger;
+
+public final class JobPollerImpl implements JobPoller {
```
Carbon copy of JobPoller just renamed. No changes.

npepinpe commented on clients/java/src/main/java/io/camunda/zeebe/client/impl/worker/JobWorkerImpl.java:
```diff
@@ -219,13 +223,24 @@ private void backoff(final JobPoller jobPoller, final Throwable error) {
   }
 
   private void handleJob(final ActivatedJob job) {
+    metrics.jobActivated(1);
```
:thought_balloon: There's some clear duplication with the handlers here, primarily the choice of the done callback. I wanted to keep the changes as simple as possible for now, but if you'd like we can tackle it here immediately.

npepinpe commented:
/cc @korthout - might be interesting for you to see how the Micrometer approach looks like.

Note that I did not go for the high level Observation API. I think that would be an interesting improvement to do here, as it would also add tracing into the mix "for free" (I mean, sort of but nothing is ever free :smile:). That would produce a timer to measure how long a job is received until it's handled, then we can add events (which are represented via counters) for the activated/handled part. The whole will also use tracing if enabled. 

I didn't go for it yet because:

- We can still migrate to it later (since everything is an interface for now)
- With the custom interface approach, tracing may not always be a compatible API, and it just complicates things to have this "leak" out
- It's fairly new, I think, popularized with Spring Boot 3.0; no idea how adoption is

koevskinikola commented:
> Assigning you as this is in the scope of job push, but this was sort of discussed while you were gone, so let me know if you need more context, or would rather someone else reviews this.

@npepinpe in the ZPA team we discussed that @korthout would take over the job metrics topic so I can focus more on multi-tenancy.

stefvic commented:
Hi Team,
Would you consider adding one more count metric for job poll failures like `zeebe.client.worker.job.poll.failures`? 
As for now, the JobPoller does only warn log and do exponential backoff retries defined in `JobWorkerImpl#onPollError(..)`. 
That kind of metric will be useful on detecting any network issues on client worker implementations  and writing corresponding  Prometheus alerting rules.

Here's how the logs look like in case of any network issues:
```
WARN io.camunda.zeebe.client.impl.worker.JobPoller - Failed to activate jobs for worker {worker_name} and job type {job_type}

io.grpc.StatusRuntimeException: UNAVAILABLE: Network closed for unknown reason
	at io.grpc.Status.asRuntimeException(Status.java:535)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:487)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:563)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:744)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)

or 

WARN io.camunda.zeebe.client.impl.worker.JobPoller - Failed to activate jobs for worker {worker_name} and job type {job_type}

io.grpc.StatusRuntimeException: UNAVAILABLE: Failed ALPN negotiation: Unable to find compatible protocol
Channel Pipeline: [SslHandler#0, ProtocolNegotiators$ClientTlsHandler#0, WriteBufferingAndExceptionHandler#0, DefaultChannelPipeline$TailContext#0]
	at io.grpc.Status.asRuntimeException(Status.java:535)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:487)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:563)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:744)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

npepinpe commented:
Hey @stefvic, that sounds reasonable eventually. For now, I'd rather we start small and build out from there. As a workaround, you can already add metrics for the job poller errors by using interceptors for your client.

You can use `ZeebeClientBuilder#withInterceptors`, and pass an interceptor like [this Prometheus one](https://github.com/grpc-ecosystem/java-grpc-prometheus/blob/master/src/main/java/me/dinowernli/grpc/prometheus/MonitoringClientInterceptor.java), which will add a metric with a status code label for every request. You can then monitor the `ActivateJobsRequest` and their status, as they will mostly come from the job poller.

korthout reviewed (changes requested):
👍 I really enjoyed reviewing this, well done @npepinpe 👏 

👍 The choice of Micrometer over OpenTelemetry seems right to me. The main argument being the existence of a bridge from Micrometer to OpenTelemetry, but not the other way.

💭 I would also be fine with a more general Micrometer facade [as you described in Slack](https://camunda.slack.com/archives/C037W9NMATG/p1691154984667949?thread_ts=1690790145.290519&cid=C037W9NMATG), although I don't yet know what benefit that would provide over the current solution with the custom interface and the default Micrometer implementation. The current solution is very flexible but perhaps more maintenance-heavy. I'd love to hear your thoughts about this.

🔧 I was wondering whether we're missing some default tags (e.g. `jobWorkerName` and `jobType`) but we can also scope this out for now.

❌ In addition, I have a few suggestions and questions. My main concern is that some tests may not be testing the job polling but instead test the job streaming. Please have a look.

💭 PS: I liked the comments in the pull request, as they were helpful. They would also have been useful in the commit messages where they could have a longer-lasting effect.

npepinpe commented on clients/java/src/main/java/io/camunda/zeebe/client/impl/worker/JobWorkerImpl.java:
```diff
@@ -100,7 +104,7 @@ public JobWorkerImpl(
   }
 
   private void openStream() {
-    jobStreamer.openStreamer(this::handleJob);
+    jobStreamer.openStreamer(this::handleStreamedJob);
```
Yes, see the linked issue: https://github.com/camunda/zeebe/issues/13814

It could be split off, but it was necessary for this feature to work properly, and I think small enough that it's fine to include here.

npepinpe commented:
> 💭 I would also be fine with a more general Micrometer facade [as you described in Slack](https://camunda.slack.com/archives/C037W9NMATG/p1691154984667949?thread_ts=1690790145.290519&cid=C037W9NMATG), although I don't yet know what benefit that would provide over the current solution with the custom interface and the default Micrometer implementation. The current solution is very flexible but perhaps more maintenance-heavy. I'd love to hear your thoughts about this.

I think a general client Micrometer facade would be interesting, but I don't have enough user data about its adoption to know whether it's worth coupling ourselves to it (like with SLF4J), or we're better off with a custom interface + optional implementation. If we did couple ourselves to Micrometer, then I would propose doing it overall for the client, and not only on the job worker. As you mentioned, keeping a custom API on top is definitely more maintenance, and it does feel like reinventing the wheel a little bit.

> 🔧 I was wondering whether we're missing some default tags (e.g. jobWorkerName and jobType) but we can also scope this out for now.

I thought about that, but these are potentially high cardinality tags. I wasn't sure if every user would want them. It's possible to add them manually, e.g.

```java
final var tags = Tags.of("com.acme.client.worker", "myWorker").and("com.acme.client.jobType", "type");
final var metrics = JobWorkerMetrics.micrometer().withTags(tags).build();
final var worker = client.newJobWorker()
    .jobType("type")
    .handler(...)
    .name("worker")
    .metrics(metrics)
    .open();
```

I suspect you're right and most users will add them. It's still something we can add in the future, but I'm also fine adding them now. It would make the builders a bit more complicated however, since somehow the metrics builder/implementation needs to be aware of the job worker properties.

> 💭 PS: I liked the comments in the pull request, as they were helpful. They would also have been useful in the commit messages where they could have a longer-lasting effect.

I'm never sure if people read the commits, but I know for sure they read the comments here, so... :sweat_smile:

npepinpe commented on clients/java/src/main/java/io/camunda/zeebe/client/impl/worker/JobWorkerBuilderImpl.java:
```diff
@@ -147,6 +150,12 @@ public JobWorkerBuilderStep3 streamTimeout(final Duration timeout) {
 
+  @Override
+  public JobWorkerBuilderStep3 metrics(final JobWorkerMetrics metrics) {
+    this.metrics = Optional.ofNullable(metrics).orElse(JobWorkerMetrics.noop());
```
That's valid, I have no idea why I ended up doing that here :shrug:

npepinpe commented on clients/java/src/test/java/io/camunda/zeebe/client/impl/worker/metrics/MicrometerJobWorkerMetricsTest.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Override
+    public boolean matches(final MeterRegistry value) {
+      final Counter counter = value.find(name.asString()).tags(tags).counter();
+      if (counter == null) {
+        return false;
+      }
+
+      return counter.count() == count;
```
I actually got more familiar with conditions recently (see the `JobStreamServiceAssert` and `JobStreamClientAssert`, for example), so I have a better idea to refactor this I think.

npepinpe commented on clients/java/src/test/java/io/camunda/zeebe/client/impl/worker/JobWorkerMetricsTest.java:
```diff
@@ -0,0 +1,215 @@
+    try (final JobWorkerImpl ignored = createWorker(0, poller, metrics)) {
+      // when
+      executor.tick(1, TimeUnit.MINUTES);
+      poller.streamJob();
+      poller.streamJob();
```
The stream job name comes from the fact both are gRPC streams (whether polling or not), but I can see how that's confusing :smile: What about `produceJob`? No verb is particularly good for a poller, since here we want to send a job to the poll consumer.

npepinpe replied:
> ❌ it feels to me that this test does not verify what is described by its name. Could you please check for me that this is the expected implementation? I worry that this tests the job streaming instead.

We're explicitly testing polling here because we actually invoke the JobPoller's consumer directly when producing the job. So we don't touch the `TestJobStreamer` consumer at all. Does that make sense? 

One of your comments earlier was about fixing a bug within the `JobWorker` (see https://github.com/camunda/zeebe/pull/13818#discussion_r1296072083). This means there are effectively two different consumers, even if they end up using the same factory. If you're not sure, you can debug the test and put a breakpoint on `JobWorkerImpl#handleStreamedJob` and `JobWorkerImpl#handleJob`, the two effective consumers, and see which is called.

npepinpe commented on clients/java/src/test/java/io/camunda/zeebe/client/impl/worker/JobWorkerMetricsTest.java:
```diff
@@ -0,0 +1,215 @@
+    final TestJobStreamer streamer = new TestJobStreamer();
+    final TestJobWorkerMetrics metrics = new TestJobWorkerMetrics();
+
+    try (final JobWorkerImpl ignored = createWorker(0, streamer, metrics)) {
```
Do you mean in general, or the semantics of using 0 specifically? I assume it's the latter :+1:

npepinpe commented on clients/java/src/test/java/io/camunda/zeebe/client/impl/worker/JobWorkerMetricsTest.java:
```diff
@@ -0,0 +1,215 @@
+import org.jmock.lib.concurrent.DeterministicScheduler;
+import org.junit.jupiter.api.Test;
+
+final class JobWorkerMetricsTest {
```
> 🔧 Given the simple nature of these tests, you could consider removing some duplication. However, I'm fine with the current duplication if that keeps the tests easier to understand

My first approach was to parametrize them, but it seemed needlessly complex. But maybe I didn't have the right approach, so if you have some ideas, I'm happy to include them here. I think a better solution would be revising the way the metrics are handled in `JobWorkerImpl`, but I think that can be improved later. Then perhaps the duplication in the tests can be removed.

npepinpe commented on clients/java/src/test/java/io/camunda/zeebe/client/impl/worker/JobWorkerMetricsTest.java:
```diff
@@ -0,0 +1,215 @@
+      consumerRef.set(jobConsumer);
+    }
+
+    private void streamJob() {
+      final ActivatedJobImpl job = new ActivatedJobImpl(mapper, TestData.job());
+      Optional.ofNullable(consumerRef.get()).ifPresent(c -> c.accept(job));
+    }
```
See the comment above on why it doesn't bypass anything. However, your point about passing noop implementations to make it clearer is a good one, so I'll do that :+1: Better safe than sorry

npepinpe commented:
@korthout I think I've addressed all your points. Let me know if I forgot something, or if there's still some nits to workout (also happy to do it f2f).

korthout reviewed (approved):
Thanks @npepinpe 🚀 

> > 🔧 I was wondering whether we're missing some default tags (e.g. jobWorkerName and jobType) but we can also scope this out for now.
> 
> I thought about that, but these are potentially high cardinality tags. I wasn't sure if every user would want them. It's possible to add them manually, e.g.

💭 Let's leave these default tags out of scope for now. Users can easily define these themselves, and they might impact performance negatively.

> I hope we can get this merged in a state that makes it easy later for to extend (e.g. add more metrics, preset tags, etc.)

It looks pretty maintainable to me. Thanks for all your efforts on this 🙇 

👍 LGTM

npepinpe commented:
bors merge
